File: apps/web/src/features/admin/userAgent.ts
export type ParsedUserAgent = {
  browser: string;
  os: string;
  label: string;
};

export function parseUserAgent(userAgent?: string | null): ParsedUserAgent {
  const value = userAgent || "";
  const browser = parseBrowser(value);
  const os = parseOperatingSystem(value);
  return {
    browser,
    os,
    label: `${browser} on ${os}`
  };
}

function parseBrowser(userAgent: string) {
  if (userAgent.includes("Edg/")) {
    return "Edge";
  }
  if (userAgent.includes("Firefox/")) {
    return "Firefox";
  }
  if (userAgent.includes("Chrome/") && !userAgent.includes("Edg/")) {
    return "Chrome";
  }
  if (
    userAgent.includes("Safari/") &&
    !userAgent.includes("Chrome/") &&
    !userAgent.includes("Chromium/") &&
    !userAgent.includes("Edg/")
  ) {
    return "Safari";
  }
  return "Unknown browser";
}

function parseOperatingSystem(userAgent: string) {
  if (userAgent.includes("iPad")) {
    return "iPadOS";
  }
  if (userAgent.includes("iPhone")) {
    return "iOS";
  }
  if (userAgent.includes("Windows NT")) {
    return "Windows";
  }
  if (userAgent.includes("Mac OS X") || userAgent.includes("Macintosh")) {
    return "macOS";
  }
  if (userAgent.includes("Android")) {
    return "Android";
  }
  if (userAgent.includes("Linux")) {
    return "Linux";
  }
  return "Unknown OS";
}

File: backend-skeleton/server/internal/auth/user_agent.go
package auth

import "strings"

type parsedUserAgent struct {
	Browser string
	OS      string
	Label   string
}

func parseUserAgent(userAgent string) parsedUserAgent {
	browser := parseUserAgentBrowser(userAgent)
	os := parseUserAgentOS(userAgent)
	return parsedUserAgent{
		Browser: browser,
		OS:      os,
		Label:   browser + " on " + os,
	}
}

func deviceLabelFromUserAgent(userAgent string) *string {
	trimmed := strings.TrimSpace(userAgent)
	if trimmed == "" {
		return nil
	}
	label := parseUserAgent(trimmed).Label
	return &label
}

func parseUserAgentBrowser(userAgent string) string {
	if strings.Contains(userAgent, "Edg/") {
		return "Edge"
	}
	if strings.Contains(userAgent, "Firefox/") {
		return "Firefox"
	}
	if strings.Contains(userAgent, "Chrome/") && !strings.Contains(userAgent, "Edg/") {
		return "Chrome"
	}
	if strings.Contains(userAgent, "Safari/") &&
		!strings.Contains(userAgent, "Chrome/") &&
		!strings.Contains(userAgent, "Chromium/") &&
		!strings.Contains(userAgent, "Edg/") {
		return "Safari"
	}
	return "Unknown browser"
}

func parseUserAgentOS(userAgent string) string {
	if strings.Contains(userAgent, "iPad") {
		return "iPadOS"
	}
	if strings.Contains(userAgent, "iPhone") {
		return "iOS"
	}
	if strings.Contains(userAgent, "Windows NT") {
		return "Windows"
	}
	if strings.Contains(userAgent, "Macintosh") || strings.Contains(userAgent, "Mac OS X") {
		return "macOS"
	}
	if strings.Contains(userAgent, "Android") {
		return "Android"
	}
	if strings.Contains(userAgent, "Linux") {
		return "Linux"
	}
	return "Unknown OS"
}
